### PR TITLE
fix(termdebug): replace mapset with nvim_set_keymap

### LINKED ISSUE #15458: Termdebug: missing function: mapset
### Neovim Version

NVIM v0.5.0

### Operating system/version:

Void Linux

### Terminal name/version:

kitty 0.23.1

### TERM environment variable

xterm-kitty

### Installation

System package manager

### Steps to reproduce

Start bare neovim:
`$ nvim -u NONE`
Map K to something:
`:nnoremap K i`
Add Termdebug package and initialize a debugging session (tested with gdb):
`:packadd termdebug`
`:Termdebug`
Quit gdb:
`(gdb) quit`

### Vim

Yes, vim works just fine. Vim version: 8.2, included patches: 1-2874

### Expected behavior

Neovim should stop the debugging session without throwing an error. 

### Actual behavior

Neovim throws the following error:

     Error detected while processing function <SNR>_1EndTermDebug[3]..<SNR>1_EndDebugCommon[21]..<SNR>1_DeleteCommands:
     line 21:
     E117: Unknown function: mapset

Other than that, neovim quits the session normally.
Taking a look at the source code, I believe this behavior is due to neovim lacking such function.

### PR DESCRIPTION
**Pull request by ii14**
Fix #15458

### CODE REVIEW

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim [hunk 1]
```diff
@@ -826,7 +826,14 @@ func s:InstallCommands()
...
   if !exists('g:termdebug_map_K') || g:termdebug_map_K
-    let s:k_map_saved = maparg('K', 'n', 0, 1)
+    " let s:k_map_saved = maparg('K', 'n', 0, 1)
+    let s:k_map_saved = {}
+    for map in nvim_get_keymap('n')
+      if map.lhs ==# 'K'
+        let s:k_map_saved = map
+        break
+      endif
+    endfor
```
Comment (clason) [inline runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:836]: No need to change this, `maparg` works perfectly fine in Neovim.

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim [hunk 2]
```diff
@@ -870,7 +877,15 @@ func s:DeleteCommands()
     if empty(s:k_map_saved)
       nunmap K
     else
-      call mapset('n', 0, s:k_map_saved)
+      " call mapset('n', 0, s:k_map_saved)
+      let mode = s:k_map_saved.mode !=# ' ' ? s:k_map_saved.mode : ''
```
Comment (clason) [inline runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: Why is this needed? I don't see `mode` being used anywhere here.

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim [hunk 3]
```diff
@@ -870,7 +877,15 @@ func s:DeleteCommands()
...
+      call nvim_set_keymap('n', 'K', s:k_map_saved.rhs, {
+        \ 'expr': s:k_map_saved.expr ? v:true : v:false,
+        \ 'noremap': s:k_map_saved.noremap ? v:true : v:false,
+        \ 'nowait': s:k_map_saved.nowait ? v:true : v:false,
+        \ 'script': s:k_map_saved.script ? v:true : v:false,
+        \ 'silent': s:k_map_saved.silent ? v:true : v:false,
```
Comment (clason) [inline runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:887]: You can just pass `expr` etc. directly; the fields always exist and have the correct truthiness in vimscript.

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim (hunk 2 above)
Comment (ii14) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: Oh shoot, my bad, it supposed to be the first argument to `nvim_set_keymap`, because it takes empty string as `:map`

Comment (clason) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: I'd test how `nvim_set_keymap`  behaves if `s:k_map_saved` is "empty" (meaning `mapset` is called on an nonexisting mapping). Might be perfectly fine without extra acrobatics!

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim (hunk 1 above)
Comment (ii14) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:836]: The thing is that, as we were discussing, `maparg` returns `<SID>` and `nvim_get_keymap` does the expansion to `<SNR>X_`.

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim (hunk 3 above)
Comment (ii14) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:887]: `nvim_set_keymap` doesn't like that unfortunately, and throws an error that it expected a boolean and got a number. Even `==` operator doesn't work.

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim (hunk 1 above)
Comment (clason) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:836]: I just want to make sure that's actually a problem -- does `nvim_set_keymap` _break_ with `<SID> `?

Comment (ii14) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:836]: Yes, it tries to link it to the `termdebug.vim` script instead of where it was actually defined previously.

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim (hunk 2 above)
Comment (ii14) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: Not sure if I understand what you mean. You mean when `s:k_map_saved` is an empty dictionary?

Comment (clason) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: Yep -- which is what you're testing here, right? (You already know the only possible mode is `n `.)

Comment (ii14) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: I did this because when you're defining a mapping with `:map` and then restoring it with `:nmap` it kinda splits it into two:
```
:map K
n  K           * ...
ov K           * ...
```
While doing it this way -- checking the actual mode -- sets it back as it was:
```
:map K
   K           * ...
```
But I guess this is still the same thing, so maybe it is indeed unnecessary. And `:unmap K` still removes both mappings.

Comment (ii14) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: And the check if dictionary was empty was already there, I just didn't touch it.

Comment (clason) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: Really? The diff says it's new ;)

Comment (ii14) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: You mean the `if empty(s:k_map_saved)` line, 877? That's weird, to me it starts on `" call mapset('n', 0, s:k_map_saved)`, line 880

Comment (clason) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: Ah, sorry, I'm talking about `s:k_map_saved.mode !=# ' ' ? s:k_map_saved.mode : ''`

Comment (clason) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: Anyway, there is an actual benefit and this isn't exactly a hot path, so it's good :)

Comment (ii14) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:881]: Oh okay. So basically the issue was that `nvim_get_keymap` returns a space (`' '`) as `:map`, and `nvim_set_keymap` expects an empty string (`''`) for `:map`.

**Comment by clason** [comment]
Thanks for the fix -- and for indulging me with my questions!
